token/services/db/sql/driver/driver.go

```diff
@@ -0,0 +1,170 @@
+
+import (
+	"database/sql"
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/hyperledger-labs/fabric-smart-client/platform/view"
+	"github.com/hyperledger-labs/fabric-smart-client/platform/view/services/flogging"
+	"github.com/hyperledger-labs/fabric-token-sdk/token"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/core/config"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/auditdb"
+	auditdbd "github.com/hyperledger-labs/fabric-token-sdk/token/services/db/driver"
+	sql2 "github.com/hyperledger-labs/fabric-token-sdk/token/services/db/sql"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/tokendb"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/ttxdb"
+	"github.com/pkg/errors"
+)
+
+var logger = flogging.MustGetLogger("token-sdk.sql")
+
+const (
+	// OptsKey is the key for the opts in the config
+	OptsKey   = "unitydb.persistence.opts"
+	EnvVarKey = "UNITYDB_DATASOURCE"
+)
+
+type Opts struct {
+	Driver       string
+	DataSource   string
+	CreateSchema bool
+	MaxOpenConns int
+}
+
+type Driver struct {
+	mutex sync.RWMutex
+	dbs   map[string]*sql.DB
+}
+
+func NewDriver() *Driver {
+	return &Driver{dbs: make(map[string]*sql.DB)}
+}
+
+func (d *Driver) OpenTokenTransactionDB(sp view.ServiceProvider, tmsID token.TMSID) (auditdbd.TokenTransactionDB, error) {
+	sqlDB, opts, name, err := d.open(sp, tmsID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to open db at [%s:%s:%s]", OptsKey, EnvVarKey, opts.Driver)
+	}
+
+	return sql2.NewTransactionDB(sqlDB, "ttx", name, opts.CreateSchema)
+}
+
+func (d *Driver) OpenTokenDB(sp view.ServiceProvider, tmsID token.TMSID) (auditdbd.TokenDB, error) {
+	sqlDB, opts, name, err := d.open(sp, tmsID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to open db at [%s:%s:%s]", OptsKey, EnvVarKey, opts.Driver)
+	}
+
+	return sql2.NewTokenDB(sqlDB, "tokens", name, opts.CreateSchema)
+}
+
+func (d *Driver) OpenAuditTransactionDB(sp view.ServiceProvider, tmsID token.TMSID) (auditdbd.AuditTransactionDB, error) {
+	sqlDB, opts, name, err := d.open(sp, tmsID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to open db at [%s:%s:%s]", OptsKey, EnvVarKey, opts.Driver)
+	}
+
+	return sql2.NewTransactionDB(sqlDB, "audit", name, opts.CreateSchema)
+}
+
+func (d *Driver) open(sp view.ServiceProvider, tmsID token.TMSID) (*sql.DB, *Opts, string, error) {
+	opts := &Opts{}
+	tmsConfig, err := config.NewTokenSDK(view.GetConfigService(sp)).GetTMS(tmsID.Network, tmsID.Channel, tmsID.Namespace)
+	if err != nil {
+		return nil, nil, "", errors.WithMessagef(err, "failed to load configuration for tms [%s]", tmsID)
+	}
+	if err := tmsConfig.UnmarshalKey(OptsKey, opts); err != nil {
+		return nil, nil, "", errors.Wrapf(err, "failed getting opts for vault")
+	}
+	if opts.Driver == "" {
+		panic(fmt.Sprintf("%s.driver not set. See https://github.com/golang/go/wiki/SQLDrivers", OptsKey))
+	}
+
+	name := fmt.Sprintf("%s-%s-%s", tmsID.Network, tmsID.Channel, tmsID.Namespace)
```

Comment:
Does this need to be in this method? We could have a separate static function `DatasourceName(tmsID)` and not even call it in here.